x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -597,4 +718,16 @@ private static boolean verifyBcryptHash(SecureString text, char[] hash) {
         SECURE_RANDOM.nextBytes(salt);
         return salt;
     }
+
+    private static SecureString getPadedPassword(SecureString data) {
+        if (data.length() < PBKDF2_APPROVED_ONLY_PWD_LIMIT) {
+            final int size = data.length();
+            final char[] paddedPassword = new char[PBKDF2_APPROVED_ONLY_PWD_LIMIT];
+            for (int i = 0; i < PBKDF2_APPROVED_ONLY_PWD_LIMIT; i++) {
+                paddedPassword[i] = data.getChars()[i%size];
+            }
+            data = new SecureString(paddedPassword);
```

Comment:
Stretching done this way effectively means that there are at least two passwords that can be verified against a single entry. For example, if the user password is `123456`, it will be stretched to `12345612345612`. So you can use either of them to login. It is worse when the actual legit password is the longer version, i.e. `12345612345612`, but we allow the much shorter version, `123456`, to verified for it. In addtion, `123456123456` is also equivalent.

The chance of this happening may be low. But given how sensitive this part of data is, I suggest we should employ a more robust padding mechanism, something similar to PKCS#7 padding, essentially add padding for all passwords regardless of its length.

Reply:
I like the PKCS7 padding idea too. An alternative could be a single round of a sha family algorithm before feeding the output (hex encoded chars) to the pbkdf2.  By virtue of the properties of the hash function, it doesnt harm pre-image resistance. Even if sha (let's say sha512) could be a surjective function, a single round does not reduce the search space enough to be considered problematic, especially since a potential attacker only controls the pre-hashed input.

Reply:
I like both PKCS7 and adding a round of sha hashing before pbkdf2. The simplest however seems to be just tripling the password before feeding it to pbkdf2 regardless of the length and character set. Since we have password min length 6 (correct me if I am wrong) it guarantees to be > 14.
Any objections/thoughts?

Reply:
Overall I think I prefer hashing before pbkdf2. It feels more standardish since pbkdf2 internally uses sha512 (though pbkdf2's usage of sha512 is a bit more involved compared to ours). I agree with Ioannis's analysis that it should not cause any practical problem. A single round of sha512 reduces the search space from essentially inifinity to `2^512`. This is still a huge number. If an attacker gets access to the final hashed result from pbkdf2 of cost 1000 (lowest we support), brute force a password requires searching through `2^512` entries (assuming a rainbow table is available to lookup the actual password once its sha512 value is known). Based on @albertzaharovits's hashing benchmark, this means more than `10^141` year of computation time on GPU or roughly `10^144` dollar value with current price of Cloud offerings. Both numbers are rather impractical.

Also we probably need to document how the hashing works since the PutUser API can take externally computed hashes. In this case, an extra sha512 is straighforward for us to document and easier for users to implement. 

Padding or stretching feels more arbitrary. Even if we follow pkcs#7 standard, it would be our own implementation and add an extra thing to worry about (maintence, testing etc). Also they feel like we are designing our own crypto "related" algorithm whose implication may be unknown.

Reply:
> A single round of sha512 reduces the search space from essentially infinity to 2^512

### For exhaustive key search attacks

Keeping in mind that most use cases concern human selected passwords or randomly generated passwords of a certain length from a specific pool, it makes it even easier to argument in favor of the one hash round, granted that the search space for a 12 char password with lower/uppercase letters, numbers and special characters is 12^95 ~~= 2^79. 

### For dictionary attacks 

Nothing changes as the password dictionary would most probably contain pre-hashed common passwords.

Reply:
Added SHA-512 Message Digest and a round of hashing using it as the stretch method